### PR TITLE
Add Transformer & Dataset

### PR DESCRIPTION
**Pull request by lbaierreinio**
1) Restructure layers & models directory to the following
layers
-> rnn
-> transformer
models

2) Add Transformer model into repository
3) Add Transformer model that interleaves chunked attention w/ minGRU into repository
4) Add Experiments directory with configuration files, python scripts to generate & store the synthetic dataset, and python script to execute an experiment
5) Add helpers for training & evaluating
6) Add logic to connect to SLURM cluster
7) Add code that generates Synthetic Datasets
8) Reformat files using base settings from autopep8 on VSCode

My apologies as this PR got quite large!

### CODE REVIEW

File: src/cs-cluster/README.md
Comment (lbaierreinio) [inline src/cs-cluster/README.md:1]: Maybe best to disinclude cs-cluster directory from the REPO since it is specific to each team member's U of T cslab account

File: src/datasets/utility.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+        # Draw sequence length from beta distribution
+        cur_seq_len = min(32, int(np.random.beta(alpha, beta) * seq_len))
+
+        # Draw split of sequences from normal distribution centered around middle of sequence
+        split = np.clip(int(np.random.normal(cur_seq_len // 2,
+                        cur_seq_len * 0.05)), 8, cur_seq_len - 8)
+
+        sequence = generate_grammar(
+            grammars[order[0]], split) + generate_grammar(grammars[order[1]], cur_seq_len - split)
```
Comment (dlrudwo1269) [inline src/datasets/utility.py:57]: Question: is there a reason for using this beta distribution and sampling the other split's length in this way?
This along with how we sample indicator indices using a normal distribution centered at the middle of the sequence would create very biased examples, no?

**Review by dlrudwo1269 — CHANGES_REQUESTED** [review]
I did not look at anything under layers/transformers yet but wanted to get something out for the day

File: src/train/utility.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+def evaluate(model, dataloader, loss_fn, evaluation_type='Validation'):
+    with torch.no_grad():
+        model.eval()
+        total_loss = 0.
+        total_correct = 0.
+        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+        for batch in dataloader:
+            input = batch['input_ids'].to(device)
+            labels = batch['labels'].to(device)
+            output = model(input)
```
[the remark this answers is not in the document]
Comment (dlrudwo1269) [reply src/train/utility.py:14]: Update: I added logic for handling padding tokens for minGRU here https://github.com/lbaierreinio/min-gru-transformers/pull/18/commits/ecbdfd2b053c8612ebb0a2f351cc442768e91eec (I'm pretty sure it's correct, but another set of eyes to verify would be nice)

File: src/datasets/utility.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        sequence[idx_one] = np.random.choice(indicators)
+        sequence[idx_two] = np.random.choice(indicators)
+
+        # Compute ground truth
+        first = sequence[idx_one]
+        second = sequence[idx_two]
+        if first == second:
+            label += halved_num_labels
+
+        sequence = sequence + ['PAD'] * (seq_len - len(sequence))
```
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/datasets/utility.py:73]: Good point! However, since within this function, we are storing the sequences in a Numpy matrix with a fixed number of rows & columns, we can't insert sequences of variable length within it.

File: src/experiments/evaluate_model.py
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/experiments/evaluate_model.py:1]: This script is that so we can evaluate the model on a number of a different datasets to see how it generalizes to harder problems!

File: src/models/minGRULM.py
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/models/minGRULM.py:1]: No, I don't believe we ended up using this for anything

File: src/datasets/utility.py (hunk 1 above)
Comment (lbaierreinio) [reply src/datasets/utility.py:57]: From what I've seen, the model most likely won't be able to learn the X,Y pattern when X & Y are uniformly distributed across the sequence length unless we throw a ton of examples at the problem. Similarly, I've found that longer examples are much harder for the model to train directly on. Therefore, if we use one sequence length of 2048, with X,Y tokens anywhere in the sequence, the number of examples needed to train directly on this task and model size may become prohibitively large!

1) Beta distribution for sequence length
Note that NLP datasets typically do not have a fixed sequence length. The beta distribution skews the dataset to include more shorter examples, as these seem easier to learn.

2) Normal Distribution for Split
Previously, the two grammars were split exactly down the middle. I used this normal distribution to add a bit of variance as to how they are split for a given example.

3) Normal Distribution for Indicator Tokens
If we distribute the indicator tokens uniformly, the amount of examples we need for the model to train directly may become quite large as mentioned! Futhermore, recall that one of the challenges of long-context reasoning is ensuring that relevant data isn't at the beginning or end of the sequence - this actually enforces that behaviour as a side-effect! 


The core idea here is to train the model on a dataset that it could learn in a reasonable amount of time - setting the distribution parameters so not all examples are extremely long, and the tokens typically aren't too far apart from one another. Again, in practice, most datasets consist of examples that are easier and examples that are more difficult. After training on such a dataset, we can see how each model generalizes to longer sequences & more uniformly spread out tokens.

If one sits down and thinks about it, one realizes there isn't much to gain from training directly on a very challenging task (2048 sequence length, tokens located anywhere), as it is unclear how long this would take and with how many examples.

**Comment by lbaierreinio** [comment]
@dlrudwo1269 Thanks for the feedback! Resolved a handful of these as I believe they became outdated. Left the ones open that I'll plan to address a bit later in the day

File: src/datasets/utility.py (hunk 1 above)
Comment (dlrudwo1269) [reply src/datasets/utility.py:57]: You mentioned "The core idea here is to train the model on a dataset that it could learn in a reasonable amount of time" so I assume this will not be our final version.

But as it is now, I think we are possibly stepping into that realm of designing our task so that it works well on our model, instead of designing a task that properly evaluates the properties we want. For example, the model might (when trained well) learn to mostly ignore the first 1/4 and last 1/4 of the sequence, it may learn to only work well on the shorter sequence (since there is more of them), etc. Additionally, if we are making the comparison to a more realistic NLP task (as we're doing to say that NLP tasks typically have more easier examples, more shorter examples, etc.), those also have the the property of most of the answers appearing at the start or end. So I think we should be consistent in which approach we take.

Just food for thought.

Comment (dlrudwo1269) [reply src/datasets/utility.py:57]: How much compute have you tried throwing for the (2048 sequence length, tokens located anywhere) task?

Can we just submit a slurm job that would run for ~48 hours of something in the meantime just to gauge the feasibility of this?

Comment (lbaierreinio) [reply src/datasets/utility.py:57]: Those are good points. I guess my thought was there might be a middle ground between the (2048 sequence length, tokens located anywhere) task and 'designing a task so that it works well on our model'.

My idea was to design a training dataset that potentially scales up to 2048 sequence length but isn't necessarily entirely composed of 2048 sequence length examples. A few reasons for this:

(1) Similar to the curriculum learning idea I suggested before (which empirically yielded some results), adding some variability into the task in terms of sequence length and/or token distance might be conducive to the model learning the longer sequence length relationship better. I don't think we need to handicap the model by exclusively training it on the 'hard case' (2048, tokens anywhere) if it is effective to also include 'easier' examples in the dataset so that the model can learn the relationship.

(2) I don't have data to back this up yet but my intuition is if the sequence length is varied it's possible we'd see some indication that the model is learning the task on shorter sequence lengths (e.g. validation loss exceeding 0.5 but not necessarily reaching 1.0). I believe exclusively training on (2048 sequence length, tokens located anywhere) would have that property where the validation loss is either 0.5 or shoots up to 1.0 rapidly after a certain number of epochs (or not at all within our time limits). If this intuition is correct, then even if the model isn't able to learn the (2048 sequence length, tokens located anywhere) in a timely manner, we would get some indication of the sequence length it can learn, and be able to make a better comparison between the Transformer & minGRU based on their respective validation losses on a varied-sequence length task.

(3) There is nothing stopping us from training the model on a training dataset that is partially or mostly consisting of 'easier' examples, and validating the model on (2048 sequence length, tokens located anywhere) task to see if it can generalize to the longer task. In the spirit of this curriculum learning idea, we could even do a longer 'pre-training' of the model on the slightly easier task, and a shorter fine-tuning on the (2048 sequence length, tokens located anywhere) before also validating it on that task. 

I think (2) and (3) would protect us against a scenario where neither model is able to learn the (2048, tokens located anywhere) directly, as I'm not sure what our analysis between the two would be in this case. I completely agree that we need to evaluate the model on the 'hard' task for our analysis, but I don't think that implies we need to train it exclusively on the (2048 sequence length, tokens located anywhere) task.

I've been giving this a lot of thought in the last week or so, so open to counter-arguments and your suggestions! It's a bit confusing to think through.

File: src/train/utility.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        for batch in train_dataloader:
+            input = batch['input_ids'].to(device)
+            labels = batch['labels'].to(device)
+
+            start = time.time()  # TODO: Verify this is a good way to measure time b/c batch size might not necessarily always be the same(?)
+            optimizer.zero_grad()
+            output = model(input)
+            loss = loss_fn(output, labels)
+            loss.backward()
+            total_loss += loss.item()
```
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/train/utility.py:48]: Yeah I used to have a clause that did early stopping if the `total_loss` dipped below x, but I removed it at some point. I'll just get rid of it for now & we can add it back if we need it

File: src/models/MinGRUClassifier.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        self.rnn = RNN(embedding_dim=embedding_dim, hidden_dim=self.hidden_dim,
+                       num_layers=num_layers, bidirectional=bidirectional)
+
+        # Classifier head
+        self.linear = nn.Linear(self.hidden_dim, num_classes)
+
+    def forward(self, x):
+        x = self.embedding(x)
+        x = self.rnn(x)
+        x = self.linear(x[:, -1])
```
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/models/MinGRUClassifier.py:23]: This is on my list of things to do. The one piece of nuance is that for the chunked Transformer, for fairness I'd assume we would want to modify the dataset for the Transformer s.t. we append a [CLS] token to the end of each 'chunk'.

Made an issue for this & we can address it as part of a future PR

File: src/experiments/mingru_config.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+
+@dataclass
+class MinGRUConfig:
+    """
+    Configuration for minGRU model.
+    """
+    name: str = 'mingru'
+    embedding_dim: int = 128
+    expansion_factor: float = 1.5
```
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/experiments/mingru_config.py]: Yes, we can do that instead. I saw another implementation that used an expansion factor, which is why I used one here. I had what you suggested a couple of weeks ago, but updated it to have expansion_factor to be more consistent with what I had seen online. For our purposes, probably nice to have more flexibility

File: src/train/utility.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    total_time = 0
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    for epoch in range(0, num_epochs+1):
+        model.train()
+        total_loss = 0.0
+        for batch in train_dataloader:
+            input = batch['input_ids'].to(device)
+            labels = batch['labels'].to(device)
+
+            start = time.time()  # TODO: Verify this is a good way to measure time b/c batch size might not necessarily always be the same(?)
```
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/train/utility.py:43]: @dlrudwo1269 Thanks for sharing that article! Based on what I read, it looks like we can do a bit better than using torch.cuda.synchronize() as that will also count the time required to launch the kernel (article suggests using CUDA events, which insert timestamps into the CUDA stream alongside the other kernels). I'll leave that to a future PR though

File: src/experiments/evaluate_model.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+    # (2) Load Dataset
+    dataset_config = DatasetConfig()
+
+    train_dataset = torch.load(validation_dataset_path)
+    train_dataloader = DataLoader(train_dataset, batch_size=32)
+
+    # (4) Define Model and Configuration
+
+    model = torch.load(model_in_path).cuda()
```
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/experiments/evaluate_model.py:40]: For consistency w/ the train script, I believe we should `device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')` & `.to(device)` no? Also - not sure what you mean by `Since the data is being placed on the tensor in evaluate, can we do the same for model?`

File: src/layers/rnn/RNN.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+from layers.rnn.BiMinGRU import BiMinGRU
+
+
+class RNN(nn.Module):
+    def __init__(self, *, embedding_dim, hidden_dim, num_layers=1, bidirectional=False):
+        super().__init__()
+
+        assert num_layers > 0, "Number of layers must be greater than 0"
+
+        self.embedding_rnn = (BiMinGRU(embedding_dim, hidden_dim)
```
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/layers/rnn/RNN.py:12]: After some thought I'm pretty sure having multiple dimensions in the minGRU was an artifact of when I had an additional linear layer within the minGRU class (to have the same effect as your model which has linear layers after each RNN cell). I think we only need one argument to the (Bi)MinGRU class, which is the hidden_dim parameter. I'll make this update in my branch, but I think it will also require you to change how you're calling the (Bi)MinGRU classes

File: src/train/utility.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/train/utility.py:14]: Yeah, we will need to add this for the Transformer as well. Going to think this through a bit

File: src/experiments/evaluate_model.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dlrudwo1269) [reply src/experiments/evaluate_model.py]: I meant do `.to(device)` for the model and data in the same place. But not a big deal

File: src/datasets/utility.py (hunk 1 above)
Comment (lbaierreinio) [reply src/datasets/utility.py]: @dlrudwo1269 Anyway unless you think there is any merit to the above tomorrow I'll try to get everything set up to just start training directly on the 2048 sequence length task & we can report whatever results we get from those. There becomes a point where we can't keep discussing what task to run the experiments on lol

File: src/layers/transformer/LongTransformerEncoder.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        x_out = x_chunks
+
+        for layer in self.layers:
+            x_out = layer(x_out)
+
+        x_res = x_out[:, -1, :]  # (N * B, H) Extract last token's hidden state
+        x_res = x_res.view(num_chunks, batch_size, num_hiddens)  # (N, B, H)
+        x_res = x_res.transpose(0, 1)  # (B, N, H)
+
+        x, _ = self.rnn_out(x)  # (B, N, H)
```
Comment (lbaierreinio) [inline src/layers/transformer/LongTransformerEncoder.py:46]: Just noting some changes here that I plan to make tomorrow:
(a) insert the CLS token into the transformer dataset and retrieve the [CLS] token for each chunk
(b) one the below:
- Use a linear layer from (4*embedding_dim => embedding_dim) that takes the CLS token from each attention block <- this would assume there are always only 4 chunks
- Use an attention mechanism on each of the CLS tokens to aggregate the results (this would work unless the sequence length exceeds something like 512^2)

Comment (dlrudwo1269) [reply src/layers/transformer/LongTransformerEncoder.py:46]: Sounds good. For our purposes, if we are only using this on this one task, it seems fine to pad up to a fixed max length and assume a fixed number of chunks (if that's easier), but either option sounds reasonable.

**Review by dlrudwo1269 — APPROVED** [review]
